cli/src/Api.ts

```diff
@@ -232,24 +272,174 @@ export default class Api {
+        return this.singleLinkageResult<Application>(result);
+    }
+
+    async workerApplicationById(group: WorkingGroups, workerApplicationId: number): Promise<WorkerApplication> {
+        const nextAppId = await this.workingGroupApiQuery(group).nextWorkerApplicationId() as WorkerApplicationId;
+
+        if (workerApplicationId < 0 || workerApplicationId >= nextAppId.toNumber()) {
+            throw new CLIError('Invalid worker application ID!');
+        }
+
+        return this.singleLinkageResult<WorkerApplication>(
+            await this.workingGroupApiQuery(group).workerApplicationById(workerApplicationId) as LinkageResult
+        );
+    }
+
+    protected async parseApplication(workerApplicationId: number, workerApplication: WorkerApplication) {
+        const appId = workerApplication.application_id;
+        const application = await this.hiringApplicationById(appId);
+
+        const { active_role_staking_id: roleStakingId, active_application_staking_id: appStakingId } = application;
+
+        return {
+            workerApplicationId,
+            applicationId: appId.toNumber(),
+            member: await this.memberProfileById(workerApplication.member_id),
+            roleAccout: workerApplication.role_account,
+            stakes: {
+                application: appStakingId.isSome ? (await this.stakeValue(appStakingId.unwrap())).toNumber() : 0,
+                role: roleStakingId.isSome ? (await this.stakeValue(roleStakingId.unwrap())).toNumber() : 0
+            },
+            humanReadableText: application.human_readable_text.toString(),
+            stage: application.stage.type as ApplicationStageKeys
+        };
+    }
+
+    async groupApplication(group: WorkingGroups, workerApplicationId: number): Promise<GroupApplication> {
+        const workerApplication = await this.workerApplicationById(group, workerApplicationId);
+        return await this.parseApplication(workerApplicationId, workerApplication);
+    }
+
+    protected async groupOpeningApplications(group: WorkingGroups, workerOpeningId: number): Promise<GroupApplication[]> {
+        const applications: GroupApplication[] = [];
+
+        const nextAppId = await this.workingGroupApiQuery(group).nextWorkerApplicationId() as WorkerApplicationId;
+        for (let i = 0; i < nextAppId.toNumber(); i++) {
+            const workerApplication = await this.workerApplicationById(group, i);
+            if (workerApplication.worker_opening_id.toNumber() !== workerOpeningId) {
+                continue;
+            }
+            applications.push(await this.parseApplication(i, workerApplication));
+        }
+
+
+        return applications;
+    }
+
+    async groupOpening(group: WorkingGroups, workerOpeningId: number): Promise<GroupOpening> {
+        const nextId = ((await this.workingGroupApiQuery(group).nextWorkerOpeningId()) as WorkerOpeningId).toNumber();
+
+        if (workerOpeningId < 0 || workerOpeningId >= nextId) {
```

Comment:
I've used this strategy as well and as a guard it works well as longs as the map has all keys in the range.
This is not the case for the WorkerById map for example. For that map I had to deal with it differently in the storage-node [package](https://github.com/Joystream/joystream/blob/a44d5beb011c7bb75e474b7644873fb434ac03aa/storage-node/packages/runtime-api/workers.js#L84)

Basically we need a way to check if a key exists in a map.
I've seen you in pioneer code base rely on `.isEmpty` property for example for a linked value to determine if the value actually existed in the map. That works perfectly for the value type in that case, but it can't really work in general if indeed a struct type will all fields with the "zero" value is actually a valid value.

So we should probably have a common interface/helper methods for reading map, linked_map, double_map state